skrub/_agg_joiner.py

```diff
@@ -0,0 +1,597 @@
+from sklearn.utils.multiclass import type_of_target
+from sklearn.utils.validation import check_is_fitted
+
+from skrub._utils import atleast_1d_or_none, atleast_2d_or_none
+from skrub.dataframe import DataFrameLike, SeriesLike
+from skrub.dataframe._namespace import get_df_namespace
+
+NUM_OPERATIONS = ["sum", "mean", "std", "min", "max", "hist", "value_counts"]
+CATEG_OPERATIONS = ["mode", "count", "value_counts"]
+ALL_OPS = NUM_OPERATIONS + CATEG_OPERATIONS
+
+
+def split_num_categ_operations(operations: list[str]) -> tuple[list[str], list[str]]:
+    """Separate aggregagor operators input by their type.
+
+    Parameters
+    ----------
+    operations : list of str
+        The input operators names.
+
+    Returns
+    -------
+    num_operations, categ_operations : Tuple of List of str
+        List of operator names
+    """
+    num_operations, categ_operations = [], []
+    for operation in operations:
+        # hist(5) -> hist
+        op_root = operation.split("(")[0]
+        if op_root in NUM_OPERATIONS:
+            num_operations.append(operation)
+        if op_root in CATEG_OPERATIONS:
+            categ_operations.append(operation)
+        if op_root not in ALL_OPS:
+            raise ValueError(f"operations options are {ALL_OPS}, got: {operation=!r}.")
+
+    return num_operations, categ_operations
+
+
+def check_missing_columns(
+    X: DataFrameLike,
+    columns: list[str],
+    error_msg: str,
+) -> None:
+    """All elements of main_key must belong to the columns of X.
+
+    Parameters
+    ----------
+    X : DataFrameLike
+        Input data.
+
+    main_key : list of string
+        Key used to perform join on X.
+    """
+    missing_cols = set(columns) - set(X.columns)
+    if len(missing_cols) > 0:
+        raise ValueError(error_msg)
+    return
+
+
+class AggJoiner(BaseEstimator, TransformerMixin):
+    """Aggregate auxiliary dataframes before joining them on a base dataframe.
+
+    Apply numerical and categorical aggregation operations on the columns
+    to aggregate, selected by dtypes. See the list of supported operations
+    at the parameter `agg_ops`.
+
+    The grouping columns used during the aggregation are the columns used
+    as keys for joining.
+
+    Accepts :obj:`pandas.DataFrame` and :obj:`polars.DataFrame` inputs.
+
+    Parameters
+    ----------
+    table : DataFrameLike or str or iterable
+        Auxiliary dataframe to aggregate then join on the base table.
+        The placeholder string "X" can be provided to perform
+        self-aggregation on the input data.
+
+    foreign_key : str or iterable of str
+        Select the columns from the auxiliary dataframe to use as keys during
+        the join operation.
+
+    main_key : str or iterable of str
+        Select the columns from the main table to use as keys during
+        the join operation.
+        If main_key is a list, we will perform a multi-column join.
+
+    cols : str or iterable of str, default=None
```

Comment:
also, I think this and operation, in addition to str and iterable of str, can be iterable of iterable of str

Reply:
I suggested `agg_cols` previously, @gvaroquaux prefered `cols`. I don't have a strong opinion on this, although cols being shorter is handy for data scientists working on notebooks.

Reply:
> also, I think this and operation, in addition to str and iterable of str, can be iterable of iterable of str

This PR is a first iteration, let's keep things simple for now.

Reply:
I meant the code is made to accept iterable of iterable of str, but the docstring  and type annotation don't mention it

Reply:
Oh yes correct, sorry read too fast

Reply:
actually do we really need this parameter? specifying a `cols` argument is not less work than selecting those same columns (or dropping the rest) from the aux table before passing it to the AggJoiner

Reply:
Yes we do: long term we want to enable covering the 90% cases without writing pandas (or polars) code. One reason is because long term we will enable working directly from data sources (eg directly from the database) and we want to make it easy for the user to switch across ways of working. Another reason is to supervise those choices.